terminal/bundles/org.eclipse.terminal.view.ui/src/org/eclipse/terminal/view/ui/internal/TerminalService.java

```diff
@@ -193,12 +193,21 @@ protected final CompletableFuture<?> executeServiceOperation(final Map<String, O
 
 	private CompletableFuture<?> executeServiceOperation(final TerminalServiceRunnable runnable, TerminalViewId tvid,
 			final String title, final ITerminalConnector connector, final Object data) {
-		// Execute the operation
-		if (runnable.isExecuteAsync()) {
-			return CompletableFuture.supplyAsync(() -> runnable.run(tvid, title, connector, data),
-					Display.getDefault());
-		} else {
-			return CompletableFuture.completedFuture(runnable.run(tvid, title, connector, data));
+		try { // Execute the operation
+			if (runnable.isExecuteAsync()) {
+				return CompletableFuture.supplyAsync(() -> runnable.run(tvid, title, connector, data),
+						/*
+						 * A special executor is used here because the default Display executor will
+						 * do a syncExec here, but isExecuteAsync is true, so we need to return from
+						 * executeServiceOperation *before* we execute the runnable.
+						 */
```

Comment:
I think this comment is here a bit misleading because it depends a bit on what thread calls this.

e.g. assume I call it from a "normal" thread, there is no guarantee that it executes *after* (or _before_) `executeServiceOperation` returns.

So I would simply explain here that `isExecuteAsync` means it should always be executed async in the UI thread.